src/model_ode.cpp

```diff
@@ -89,29 +90,45 @@ void Model_ODE::fJDiag(const realtype t, AmiVector &JDiag,
 void Model_ODE::fdxdotdw(const realtype t, const N_Vector x) {
     dxdotdw.reset();
     auto x_pos = computeX_pos(x);
-    fdxdotdw(dxdotdw.data(), t, N_VGetArrayPointer(x_pos),
-             unscaledParameters.data(), fixedParameters.data(), h.data(),
-             w.data());
-    fdxdotdw_colptrs(dxdotdw.indexptrs());
-    fdxdotdw_rowvals(dxdotdw.indexvals());
+    if (nw > 0) {
+        fdxdotdw_colptrs(dxdotdw.indexptrs());
+        fdxdotdw_rowvals(dxdotdw.indexvals());
+        fdxdotdw(dxdotdw.data(), t, N_VGetArrayPointer(x_pos),
+                 unscaledParameters.data(), fixedParameters.data(), h.data(),
+                 w.data());
+    }
 }
 
 void Model_ODE::fdxdotdp(const realtype t, const N_Vector x) {
     auto x_pos = computeX_pos(x);
-    fdwdp(t, N_VGetArrayPointer(x));
-    if (wasPythonGenerated()) {
+    fdwdp(t, N_VGetArrayPointer(x_pos));
+    fdxdotdw(t, x_pos);
+    
+    if (pythonGenerated) {
         // python generated
-        fdxdotdw(t, x);
-        for (int ip = 0; ip < nplist(); ip++) {
-            N_VConst(0.0, dxdotdp.getNVector(ip));
-            fdxdotdp(dxdotdp.data(ip), t, N_VGetArrayPointer(x_pos),
-                     unscaledParameters.data(), fixedParameters.data(),
-                     h.data(), plist_[ip], w.data());
-            if (nw > 0)
-                dxdotdw.multiply(
-                    gsl::span<realtype>(dxdotdp.data(ip), nx_solver),
-                    gsl::span<const realtype>(&dwdp.at(nw * ip), nw));
+        if (ndxdotdp_explicit > 0) {
+            dxdotdp_explicit.reset();
+            fdxdotdp_explicit_colptrs(dxdotdp_explicit.indexptrs());
+            fdxdotdp_explicit_rowvals(dxdotdp_explicit.indexvals());
+            
+            for (int ip = 0; ip < nplist(); ip++)
+                /* The creation of the according file has to be changed */
+                fdxdotdp_explicit(dxdotdp_explicit.data(),
+                                  t, N_VGetArrayPointer(x_pos),
+                                  unscaledParameters.data(),
+                                  fixedParameters.data(),
+                                  h.data(), plist_[ip], w.data());
+        }
+        
+        if (nw > 0 && ndxdotdp_implicit > 0) {
+            /* Sparse matrix multiplication
+               dxdotdp_implicit += dxdotdw * dwdp */
+            dxdotdp_implicit.reset();
+            fdxdotdp_implicit_colptrs(dxdotdp_implicit.indexptrs());
+            fdxdotdp_implicit_rowvals(dxdotdp_implicit.indexvals());
+            dxdotdw.sparse_multiply(&dxdotdp_implicit, &dwdp);
```

Comment:
here we no longer account for the sensitivity indexing in plist. Looking at the indexing in `dwdp` in `fdydp` and the implementation of `dwdp`, `fdydp` also seems to expect the indexed/reordered matrix, while `dwdp` seems to implement the full matrix. There are two ways to fix this.

A) implement `dwdp` to only compute plist-indexed `dwdp`. This is somewhat unpleasant because `ndwdp`, `fdwdp_colptrs` and `fdwdp_rowvals` all become plist dependent. Same will applies to `n_dxdotdp_implicit`, `fdxdotdp_implicit_colptrs` and `fdxdotdp_implicit_rowvals`.

B) change indexing in `dydp` to account for `plist` (easy) and change `sparse_multiply` to also account for plist (this is less easy). This effectively boils down to the same problem as with A), that `n_dxdotdp_implicit`, `fdxdotdp_implicit_colptrs` and `fdxdotdp_implicit_rowvals` become parameter dependent.

Reply:
Alternative C) which seems substantially easier:
Compute `dxdotdp` for all parameters, apply plist on subsequent steps. Adds some overhead, but should be negligible in most practical situations as we usually always compute full sensitivities.

Reply:
Well, we don't do that any more in dxdotdp. Instead, the accounting for plist has moved to qBdot.

Reply:
My plan here was:

> Alternative C) which seems substantially easier:
> Compute `dxdotdp` for all parameters, apply plist on subsequent steps. Adds some overhead, but should be negligible in most practical situations as we usually always compute full sensitivities.

This was exactly the way I did it for most of the things, for exactly the reasons you pointed out here. I also thought of it while implementing and came to the same conclusion... ;)

My idea was:
dwdp, dxdotdp_explicit, and dxdotdp_implicit aer always full. Applying plist only happens subsequently. So, this should happen in:
fqBdot (there it is already implemented by introducing a new multiply function, which is basically doing the same thing as the old one, but has three differences:
 * It uses only sparse CSC_matrices
 * It takes plist, reorders and only applies it
 * It can at the same time transpose matrices, which is helpful for qBdot

However, I wasn't yet sure ho to apply  plist for dydp. Will that now.